File: Interviews/LeetCode/Topic-Wise/Trees/Binary-Trees/Easy/Two-Sum-IV-Input-is-a-BST/653.cpp
class Iterator {
    vector<TreeNode*> left, right;
public:
    Iterator() {
        left.clear();
        right.clear();
    }

    int element(bool nextOrBefore) { // nextOrBefore = 0 means left iterator, else right iterator
        TreeNode* getElement;
        if (nextOrBefore) {
            if (right.size()) {
                getElement = right.back();
                right.pop_back();
                pushIntoStack(getElement->left, 1);
            }
        } else {
            if (left.size()) {
                getElement = left.back();
                left.pop_back();
                pushIntoStack(getElement->right, 0);
            }
        }

        return getElement->val;
    }

    void pushIntoStack(TreeNode* root, bool pushDir) { // pushDir = 0 means left, else right
        if (pushDir) {
            while (root) {
                right.push_back(root);
                root = root->right;
            }
        } else {
            while (root) {
                left.push_back(root);
                root = root->left;
            }
        }
    }

};

class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        Iterator obj;
        obj.pushIntoStack(root, 0), obj.pushIntoStack(root, 1);
        int l = obj.element(0), r = obj.element(1);
        while (l < r) {
            int sum = l + r;
            if (sum == k) return true;
            if (sum < k) l = obj.element(0);
            else r = obj.element(1);
        }

        return false;
    }
};

File: Interviews/LeetCode/Daily-Problems/2022/October/Day-09/653.py
class Iterator:
    def __init__(self, __reverse, root):
        self.stInc, self.stDec, self.reverse = [], [], __reverse
        node = root
        if __reverse:
            while node:
                self.stDec.append(node)
                node = node.right
        else:
            while node:
                self.stInc.append(node)
                node = node.left

    def getNext(self) -> int:
        if self.reverse:
            ans, node = self.stDec[-1], self.stDec[-1].left
            self.stDec.pop()
            while node:
                self.stDec.append(node)
                node = node.right
            return ans.val
        else:
            ans, node = self.stInc[-1], self.stInc[-1].right
            self.stInc.pop()
            while node:
                self.stInc.append(node)
                node = node.left
            return ans.val
        return -1

class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        inc, dec = Iterator(False, root), Iterator(True, root)
        i, j = inc.getNext(), dec.getNext()
        while i < j:
            sum = i + j
            if sum == k: return True
            if sum < k: i = inc.getNext()
            else: j = dec.getNext()
    
        return False